### PR TITLE
send back keep-alive to slack; closes #223

### LINKED ISSUE #223: [server] Send keep-alive to slack

### PR DESCRIPTION
**Pull request by thuey**

### CODE REVIEW

File: internal/server/slack.go [hunk 1]
```diff
@@ -24,60 +26,85 @@ type slackResponse struct {
 // Slack page
 func (s *Server) handleSlack() func(w http.ResponseWriter, r *http.Request) {
 	return func(w http.ResponseWriter, r *http.Request) {
-		cfg, err := config.Get()
+		body, err := ioutil.ReadAll(r.Body)
 		if err != nil {
```
Comment (jonnymbgood) [inline internal/server/slack.go]: I think you can just defer r.Body.Close here

File: internal/server/slack.go [hunk 2]
```diff
@@ -24,60 +26,85 @@ type slackResponse struct {
...
-			fmt.Fprint(w, "Error Retrieving Config")
+			fmt.Fprint(w, "Error reading body.")
 			return
 		}
+		r.Body.Close() //  must close
+		r.Body = ioutil.NopCloser(bytes.NewBuffer(body))
```
Comment (jonnymbgood) [inline internal/server/slack.go]: ?

File: internal/server/slack.go [hunk 3]
```diff
@@ -24,60 +26,85 @@ type slackResponse struct {
...
 
-		body, err := ioutil.ReadAll(r.Body)
-		if err != nil {
-			http.Error(w, "Error reading body.", http.StatusBadRequest)
-			return
-		}
-		r.Body.Close() //  must close
-		r.Body = ioutil.NopCloser(bytes.NewBuffer(body))
-		base := "v0:" + r.Header.Get("X-Slack-Request-Timestamp") + ":" + string(body)
+		go func(r *http.Request, body []byte) {
```
Comment (jonnymbgood) [inline internal/server/slack.go]: A recommendation. Do everything but the execution outside of the go routine, and then call the goroutine with the user, args, config, and callback url. That way you don't pass the request or body to a function that gets spun off. The body parsing and marshaling is going to be quick. The part that takes time is the action.Execute call and the callback to slack. So maybe:
1) get vars from body/header
2) do verification
3) marshal/parse vars
4) respond(write the pleaseops ism out)
5) call goroutine (which is non-blocking)

File: internal/server/slack.go (hunk 2 above)
Comment (thuey) [reply internal/server/slack.go]: I need the raw body for verification. Also, r.ParsesForm reads the body. The body can't be read twice, so this is what the internet recommended. I'm guessing there's a form parser library I could give the raw body to. I'll look into that

File: internal/server/slack.go (hunk 1 above)
Comment (thuey) [reply internal/server/slack.go]: See below comment

File: internal/server/slack.go (hunk 3 above)
Comment (thuey) [reply internal/server/slack.go]: Actually, getting the config takes a while. And I need the config in order to do the verification. By "a while", I mean that relative to the amount of time it takes to perform an action like "get po users" and then return it. I started off with the config getting and verification outside of the go routine, but the user experience was weird. I would run a command like "get po users", there would be a delay of almost a second and then both the pleaseopsism and result would come back at the exact same time. Sometimes, I would actually get the action result back first

File: internal/server/slack.go (hunk 2 above)
Comment (jonnymbgood) [reply internal/server/slack.go]: Ah, makes sense.

File: internal/server/slack.go (hunk 3 above)
Comment (jonnymbgood) [reply internal/server/slack.go]: The only real issue I have with this is passing the request pointer into the go function. I'm not sure about the guarantees of reading from a request after the response has gone out and closed.  If we refactored that piece out so we get what we need and pass it into the gofunc I would be ok with that. You could create a struct to hole everything if it gets to be too many params.

**Comment by thuey** [comment]
Can do

**Pull request merged**